Funcky/Extensions/EnumerableExtensions/AverageOrNone.cs

```diff
@@ -0,0 +1,320 @@
+        /// The rules for floating point arithmetic apply see remarks for detail.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values that are used to calculate an average.</param>
+        /// <param name="selector">A transform function to apply to each element.</param>
+        /// <returns>The average of the sequence of values or None.</returns>
+        [Pure]
+        public static Option<double> AverageOrNone<TSource>(this IEnumerable<TSource> source, Func<TSource, Option<double>> selector)
+            => source.Aggregate(AverageCalculatorDouble.Empty, (calculator, element) => calculator.Add(selector(element))).Average;
+
+        /// <summary>
+        /// Computes the average of a sequence of Decimal values that are obtained by invoking a transform function on each element of the input sequence.
+        /// If the sequence is empty it returns None.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values that are used to calculate an average.</param>
+        /// <param name="selector">A transform function to apply to each element.</param>
+        /// <returns>The average of the sequence of values or None.</returns>
+        [Pure]
+        public static Option<decimal> AverageOrNone<TSource>(this IEnumerable<TSource> source, Func<TSource, decimal> selector)
+            => source.Aggregate(AverageCalculatorDecimal.Empty, (calculator, element) => calculator.Add(selector(element))).Average;
+
+        /// <summary>
+        /// Computes the average of a sequence of optional Decimal values that are obtained by invoking a transform function on each element of the input sequence.
+        /// If the sequence only consists of none or is empty it returns None.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values that are used to calculate an average.</param>
+        /// <param name="selector">A transform function to apply to each element.</param>
+        /// <returns>The average of the sequence of values or None.</returns>
+        [Pure]
+        public static Option<decimal> AverageOrNone<TSource>(this IEnumerable<TSource> source, Func<TSource, Option<decimal>> selector)
+            => source.Aggregate(AverageCalculatorDecimal.Empty, (calculator, element) => calculator.Add(selector(element))).Average;
+
+        private class AverageCalculatorDouble
+        {
+            public static readonly AverageCalculatorDouble Empty = new AverageCalculatorDouble();
+
+            private readonly int _count;
+            private readonly Option<double> _sum;
+
+            public AverageCalculatorDouble(int count = default, Option<double> sum = default)
```

Comment:
Constructor could be private (Applies to all `Calculator`s)

Reply:
Adressed in d353e26c